x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -1165,8 +1213,134 @@ protected LogicalPlan rule(Aggregate aggregate) {
             return aggregate;
         }
 
-        static String temporaryName(NamedExpression agg, AggregateFunction af) {
-            return SubstituteSurrogates.temporaryName(agg, af);
+        static String temporaryName(Expression expression, AggregateFunction af) {
+            return SubstituteSurrogates.temporaryName(expression, af);
+        }
+    }
+
+    /**
+     * Replace nested expressions over aggregates with synthetic eval post the aggregation
+     * stats a = sum(a) + min(b) by x
+     * becomes
+     * stats a1 = sum(a), a2 = min(b) by x | eval a = a1 + a2 | keep a, x
+     *
+     * Since the logic is very similar, this rule also handles duplicate aggregate functions to avoid duplicate compute
+     * stats a = min(x), b = min(x), c = count(*), d = count() by g
+     * becomes
+     * stats a = min(x), c = count(*) by g | eval b = a, d = c | keep a, b, c, d, g
+     */
+    static class ReplaceStatsAggExpressionWithEval extends OptimizerRules.OptimizerRule<Aggregate> {
```

Comment:
The core of this PR - breaks down the expression over aggs and adds an eval lazily only for the fields that have an expression over aggregate functions.

Reply:
We could improve this a bit further for a referencing case: just like we now support `| eval x = field, y = x + 1`, we could now (but don't yet) support something like: `| stats x = max(field), y = x + min(field)` -- this now fails because column `x` isn't known.

Reply:
That would be a nice little feature - raised https://github.com/elastic/elasticsearch/issues/105102 as a follow-up.